File: PROGRAMS/2/array_leader.py
arr = [16, 17, 4, 3, 5, 2]

"""
 Brute Force Method: 
"""

def array_leader_brute_force(arr: list):
    leader_array = []
    for i in range(len(arr)):
        found_leader = True
        for j in range(i+1, len(arr)):
            if arr[j] > arr[i]:
                found_leader = False
                break
        if found_leader:
            leader_array.append(arr[i])
    return leader_array


array_leader_brute_force(arr=arr)

"""
Greedy Algorithm
"""

def array_leader(arr: list):
    leader_array = []
    max_element = arr[-1]
    for i in range(len(arr)-1, -1, -1):
        if arr[i] >= max_element:
            max_element = arr[i]
            leader_array.append(max_element)
    leader_array.reverse() 
    return leader_array


array_leader(arr=arr)

File: PROGRAMS/2/array_leader.c
#include<stdio.h>

void print_array(int arr[], int size)
{
    int i;
    printf("[");
    for(i = 0; i < size; i++)
    {
        printf("%d", arr[i]);
        if(i != size - 1) printf(", ");
    }
    printf("] \n");
}

void print_array_reverse(int arr[], int size)
{
    int i;
    printf("[");
    for(i = size-1; i > -1; i--)
    {
        printf("%d", arr[i]);
        if(i != 0) printf(", ");
    }
    printf("] \n");
}

//  Brute Force Method: Approach 1

void 
brute_force_array_leader(int arr[], int size)
{   int i, j, k = 0;
    int leader_array[size];
    for(i=0; i< size; i++)
    {
        int leader = 1;
        for(j=i+1; j< size; j++)
        {
            if(arr[j] > arr[i])
            {
                leader = 0;
                break;
            }
        }
        if(leader == 1)
        {
        leader_array[k++] = arr[i];
        }
    }
    print_array(leader_array, k);
} 

void 
array_leader(int arr[], int size)
{
    int i,j=0;
    int max_element = arr[size-1];
    int leader_array[size];
    for(i=size-1; i>-1; i--)
    {
        if(arr[i] >= max_element)
        {
            leader_array[j++] = arr[i];
            max_element = arr[i];
        }
    }
    print_array_reverse(leader_array, j);
}


int main()
{
    int arr[] = {16, 17, 4, 3, 5, 2};
    int size_of_array = sizeof(arr) / sizeof(arr[0]);
    printf("Array size: \n");
    print_array(arr, size_of_array);

    printf("Brute Force Array Leader: \n");
    brute_force_array_leader(arr, size_of_array);

    printf("Optimized Greedy Approach \n");
    array_leader(arr, size_of_array);



    return 0;

}